File: BOJ/Platinum/9936. 체스로 도미노를 타자/9936.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(1000000)
INF = int(-1e9)


def cover(idx, status, cnt):
    if cnt == 0:
        return 0

    if idx >= n * 3:
        return INF

    if dp[idx][status][cnt] != INF:
        return dp[idx][status][cnt]

    if status & 1:
        dp[idx][status][cnt] = cover(idx + 1, status >> 1, cnt)
    else:
        dp[idx][status][cnt] = cover(idx + 1, status >> 1, cnt)

        if idx + 3 < n * 3:
            dp[idx][status][cnt] = max(dp[idx][status][cnt], cover(
                idx + 1, status >> 1 | 4, cnt - 1) + chess[idx] + chess[idx + 3])

        if idx % 3 != 2 and not (status & 2):
            dp[idx][status][cnt] = max(dp[idx][status][cnt], cover(
                idx + 2, status >> 2, cnt - 1) + chess[idx] + chess[idx + 1])

    return dp[idx][status][cnt]


n, k = map(int, input().split())
chess = []
for _ in range(n):
    chess.extend(list(map(int, input().split())))

dp = [[[INF] * (k + 1) for _ in range(8)] for _ in range(3 * (n + 1))]
print(cover(0, 0, k))

File: BOJ/Platinum/9936. 체스로 도미노를 타자/9936.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const int INF = -1e9;

int n, k;
vector<int> chess;
vector<vector<vector<int>>> dp;

int cover(int idx, int status, int cnt)
{
    if (cnt == 0)
        return 0;
    if (idx >= n * 3)
        return INF;

    int &ret = dp[idx][status][cnt];

    if (ret != INF)
        return ret;

    if (status & 1)
        ret = cover(idx + 1, status >> 1, cnt);
    else
    {
        ret = cover(idx + 1, status >> 1, cnt);

        if (idx + 3 < n * 3)
            ret = max(ret, cover(idx + 1, status >> 1 | 4, cnt - 1) + chess[idx] + chess[idx + 3]);

        if (idx % 3 != 2 && !(status & 2))
            ret = max(ret, cover(idx + 2, status >> 2, cnt - 1) + chess[idx] + chess[idx + 1]);
    }

    return ret;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> n >> k;
    chess.resize(n * 3);
    dp.resize(3 * (n + 1), vector<vector<int>>(8, vector<int>(k + 1, INF)));

    for (int i = 0; i < n; ++i)
    {
        for (int j = 0; j < 3; ++j)
        {
            int val;
            cin >> val;
            chess[i * 3 + j] = val;
        }
    }

    cout << cover(0, 0, k) << endl;

    return 0;
}
